File: My Bank App/ChangePasswordTableViewController.m
//
//  ChangePasswordTableViewController.m
//  My Bank App
//
//  Created by Arthur Walasek on 9/23/14.
//  Copyright (c) 2014 Callpod, Inc. All rights reserved.
//

#import "ChangePasswordTableViewController.h"
#import "KeeperTableViewCell.h"
#import <KeeperExtensionSDK/KeeperExtensionSDK.h>

@interface ChangePasswordTableViewController ()

@end

@implementation ChangePasswordTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;

#ifdef USE_KEEPER_TAB_ICON
    [self.keeperTabBtn setHidden:![[KeeperSDK sharedExtension] isAppExtensionAvailable]];
#else
    [self.keeperTabBtn setHidden:TRUE];
#endif
    
    [self.tableView setRowHeight:44.0f];
    [self.tableView.tableHeaderView setFrame:CGRectMake(self.tableView.tableHeaderView.bounds.origin.x, self.tableView.tableHeaderView.bounds.origin.y, self.tableView.tableHeaderView.bounds.size.width, 135)];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
    return 3;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    KeeperTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"keeperCell" forIndexPath:indexPath];
    if (!cell) {
        cell = [[KeeperTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"keeperCell"];
    }
    if (indexPath.row == 0) {
        cell.txtText.secureTextEntry = TRUE;
        cell.txtText.placeholder = @"Old Password";
        if (self.oldpassword) {
            cell.txtText.text = self.oldpassword;
        }
    }
    else if (indexPath.row == 1) {
        cell.txtText.secureTextEntry = TRUE;
        cell.txtText.placeholder = @"New Password";
        if (self.password1) {
            cell.txtText.text = self.password1;
        }
    }
    else if (indexPath.row == 2) {
        cell.txtText.secureTextEntry = TRUE;
        cell.txtText.placeholder = @"Confirm Password";
        if (self.password2) {
            cell.txtText.text = self.password2;
        }
    }
    
    cell.txtText.tag = indexPath.row;
    cell.delegate = self;
    cell.returnKeyType = UIReturnKeyDone;
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    
    [cell setUpCell];
    
    return cell;
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    return YES;
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    if (textField.tag == 0) {
        self.oldpassword = textField.text;
    }
    else if (textField.tag == 1) {
        self.password1 = textField.text;
    }
    else if (textField.tag == 2) {
        self.password2 = textField.text;
    }
}

#pragma mark - Actions

- (IBAction)cancelAction:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)changeAction:(id)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Password Changed" message:@"Your password has been changed" preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *dismissAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
        [self.navigationController popViewControllerAnimated:YES];
    }];
    
    [alert addAction:dismissAction];
    [self presentViewController:alert animated:YES completion:nil];
    return;
}

- (IBAction)keeperAction:(id)sender {
    [self keeperLockAction];
}

- (void)keeperLockAction {
    NSString *changedPassword = self.password1 ? : @"";
    
    // Validate that the new and confirmation passwords match.
    if (changedPassword.length > 0 && ![changedPassword isEqualToString:self.password2]) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Change Password Error" message:@"The new and the confirmation passwords must match" preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *dismissAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
            self.password1 = @"";
            self.password2 = @"";
            [self.tableView reloadData];
        }];
        
        [alert addAction:dismissAction];
        [self presentViewController:alert animated:YES completion:nil];
        return;
    }
    
    NSDictionary *loginDetails = @{
                                   AppExtensionTitleKey: @"My Bank Registration",
                                   AppExtensionPasswordKey: changedPassword,
                                   AppExtensionOldPasswordKey: self.oldpassword ? : @"",
                                   AppExtensionNotesKey: @"Saved with the My Bank app",
                                   };
    
    // Password generation options are optional, but are very handy in case you have strict rules about password lengths
    NSDictionary *passwordGenerationOptions = @{
                                                AppExtensionGeneratedPasswordMinLengthKey: @(6),
                                                AppExtensionGeneratedPasswordMaxLengthKey: @(50)
                                                };
    
    __weak typeof (self) miniMe = self;
    
    [[KeeperSDK sharedExtension] changePasswordForLoginForURLString:@"http://www.my-bank-website.com" loginDetails:loginDetails passwordGenerationOptions:passwordGenerationOptions forViewController:self sender:self completion:^(NSDictionary *loginDict, NSError *error) {
        if (!loginDict) {
            if (error.code != AppExtensionErrorCodeCancelledByUser) {
                NSLog(@"Failed to use Keeper App Extension to change password: %@", error);
            }
            return;
        }
        
        __strong typeof(self) strongMe = miniMe;
        strongMe.oldpassword = loginDict[AppExtensionOldPasswordKey];
        strongMe.password1 = loginDict[AppExtensionPasswordKey];
        strongMe.password2 = loginDict[AppExtensionPasswordKey];
        
        [self.tableView reloadData];
    }];
}

@end

File: My Bank App Swift/ChangePasswordTableViewController.swift
//
//  ChangePasswordTableViewController.swift
//  My Bank App
//
//  Created by Jesse Gatt on 10/13/14.
//  Copyright (c) 2014 Callpod, Inc. All rights reserved.
//

import UIKit
import KeeperExtensionSDK

class ChangePasswordTableViewController: UITableViewController, UITextFieldDelegate, KeeperLockActionDelegate {

    @IBOutlet weak var keeperTabBtn: UIButton!
    var oldpassword:NSString?
    var password1:NSString?
    var password2:NSString?
    override func viewDidLoad() {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
        
        #if USE_KEEPER_TAB_ICON
            keeperTabBtn.hidden = !(KeeperSDK.sharedExtension().isAppExtensionAvailable())
        #else
            keeperTabBtn.hidden = true
        #endif
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        // #warning Potentially incomplete method implementation.
        // Return the number of sections.
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete method implementation.
        // Return the number of rows in the section.
        return 3
    }

    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("keeperCell", forIndexPath: indexPath) as KeeperTableViewCell

        // Configure the cell...
        
        switch indexPath.row {
        case 0:
            cell.txtText.secureTextEntry = true
            cell.txtText.placeholder = "Old Password"
            if let oldpassword = oldpassword {
                cell.txtText.text = oldpassword
            }
        case 1:
            cell.txtText.secureTextEntry = true
            cell.txtText.placeholder = "New Password"
            if let password1 = password1 {
                cell.txtText.text = password1
            }
        case 2:
            cell.txtText.secureTextEntry = true
            cell.txtText.placeholder = "Confirm Password"
            if let password2 = password2 {
                cell.txtText.text = password2
            }
        default:
            cell.txtText.text = ""
        }
        
        
        // Configure the cell...
        
        cell.txtText.tag = indexPath.row
        cell.delegate = self
        cell.returnKeyType = UIReturnKeyType.Done
        cell.selectionStyle = UITableViewCellSelectionStyle.None
        
        cell.setUpCell()


        return cell
    }
    
    // MARK: - UITextFieldDelegate
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
    
    func textFieldDidEndEditing(textField: UITextField) {
        switch textField.tag {
        case 0:
            oldpassword = textField.text
        case 1:
            password1 = textField.text
        case 2:
            password2 = textField.text
        default:
            println("Default Case")
        }
    }
    
    
   
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */
    @IBAction func changeAction(sender: AnyObject) {
        
        
        
        let alert = UIAlertController(title: "Password Changed", message: "Your password has been changed", preferredStyle: UIAlertControllerStyle.Alert)
        
        alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
        
        self.dismissViewControllerAnimated(true, completion: nil)
        self.presentViewController(alert, animated: true, completion: nil)
        return;
    }

    @IBAction func keeperLockAction(sender: AnyObject) {
        
        let changePassword = password1 ?? ""
        
        if (changePassword.length > 0) && !(changePassword == password2) {
            let alert = UIAlertController(title: "Change Password Error", message: "The new and the confirmation passwords must match", preferredStyle: UIAlertControllerStyle.Alert)
            
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Cancel, handler: nil))
            
            password1 = ""
            password2 = ""
            self.presentViewController(alert, animated: true, completion: nil)
            return;
        }
        
        let loginDetails = [AppExtensionTitleKey: "My Bank Registration",
            AppExtensionPasswordKey: changePassword,
            AppExtensionOldPasswordKey: oldpassword ?? "",
            AppExtensionNotesKey: "Saved with the My Bank app"]
        
        let passwordGenerationOptions = [AppExtensionGeneratedPasswordMinLengthKey: 6, AppExtensionGeneratedPasswordMaxLengthKey: 50]
        
        KeeperSDK.sharedExtension().changePasswordForLoginForURLString("http://www.my-bank-website.com", loginDetails: loginDetails, passwordGenerationOptions: passwordGenerationOptions, forViewController: self, sender: sender) { (loginDict:[NSObject : AnyObject]!, error:NSError!) -> Void in
            if error.code != AppExtensionErrorCodeCancelledByUser {
                println("Failed to use Keeper App Extension to change password: \(error)")
                return
            }
            
            self.oldpassword = loginDict[AppExtensionOldPasswordKey] as? NSString
            self.password1 = loginDict[AppExtensionOldPasswordKey] as? NSString
            self.password2 = loginDict[AppExtensionOldPasswordKey] as? NSString
            
            self.tableView.reloadData()
        }

    }
}
